support-workers/src/main/scala/com/gu/support/workers/lambdas/SendAcquisitionEvent.scala

```diff
@@ -211,10 +224,18 @@ object SendAcquisitionEvent {
           if (stateAndInfo.requestInfo.accountExists) Some("REUSED_EXISTING_PAYMENT_METHOD") else None,
           if (isSixForSix(stateAndInfo)) Some("guardian-weekly-six-for-six") else None,
           if (stateAndInfo.state.analyticsInfo.isGiftPurchase) Some("gift-subscription") else None,
-          stateAndInfo.state.paymentOrRedemptionData.right.map(_ => "corporate-subscription").toOption
+          stateAndInfo.state.productTypeCreated match {
+            case _: DigitalSubscriptionCorporateRedemptionCreated => Some("corporate-subscription")
+            case _ => None
+          }
```

Comment:
all of these 2 line pattern matches can be reduced in boilerplate by adding e.g `def asDigitalSubscriptionCorporateRedemptionCreated: Option[DigitalSubscriptionCorporateRedemptionCreated]` or `def is DigitalSubscriptionCorporateRedemptionCreated: Boolean` function on the type, but there's a tradeoff in terms of clutter.
Unless you know a simpler way?

Reply:
I raised this on scala channel, and there are a few points thanks to Adam
- we are moving the product type to the type level wheras it was more at the value level before. This allows the compiler to check a lot more in terms of grouping of possible values (this was the aim of the PR).  This also means pattern match is the basic tool to get back to the value level, rather than if statements.
- in scala standard library they added a lot of built in functions to get back to the value level e.g. `.left.toOption`
- we could add similar functions here `.as[Blah]` but the question would be is that abstraction worth learning over and above the standard matching, for the benefit it gives.
- it would be possible to do something generic with ClassTag thanks to Mario (and even generic implicit class), but it may not be the best way.

Reply:
this line could be slightly simplified to but I haven't thought whether it's a good idea yet.  Maybe one for a later PR.

```suggestion
          stateAndInfo.state.productTypeCreated.collect { case _: DigitalSubscriptionCorporateRedemptionCreated => "corporate-subscription" }
```
we would also add this collect function to the trait
```scala
sealed trait ProductTypeCreated {
  def product: ProductType
  def collect[T](f: PartialFunction[ProductTypeCreated, T]): Option[T] = f.lift(this)
}
```